controllers/kuadrant_controller.go

```diff
@@ -36,24 +48,248 @@ type KuadrantReconciler struct {
+		controllerutil.AddFinalizer(kObj, kuadrantFinalizer)
+		if err := r.Update(ctx, kObj); client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	specResult, specErr := r.reconcileSpec(ctx, kObj)
+	if specErr == nil && specResult.Requeue {
+		logger.V(1).Info("Reconciling spec not finished. Requeueing.")
+		return specResult, nil
+	}
+
+	statusResult, statusErr := r.reconcileStatus(ctx, kObj, specErr)
+
+	if specErr != nil {
+		return ctrl.Result{}, specErr
+	}
+
+	if statusErr != nil {
+		return ctrl.Result{}, statusErr
+	}
 
-	// your logic here
+	if statusResult.Requeue {
+		logger.V(1).Info("Reconciling status not finished. Requeueing.")
+		return statusResult, nil
+	}
 
+	logger.Info("successfully reconciled")
 	return ctrl.Result{}, nil
 }
 
+func (r *KuadrantReconciler) unregisterExternalAuthorizer(ctx context.Context) error {
+	logger := logr.FromContext(ctx)
+	iop := &iopv1alpha1.IstioOperator{}
+
+	iopKey := client.ObjectKey{Name: iopName(), Namespace: iopNamespace()}
+	if err := r.Get(ctx, iopKey, iop); err != nil {
+		// It should exists, NotFound also considered as error
+		logger.Error(err, "failed to get istiooperator object", "key", iopKey)
+		return err
+	}
+
+	if !hasKuadrantAuthorizer(iop) {
+		return nil
+	}
+
+	obj, ok := iop.Spec.MeshConfig["extensionProviders"]
+	if !ok || obj == nil {
+		obj = make([]interface{}, 0)
+	}
+
+	extensionProviders, ok := obj.([]interface{})
+	if !ok {
+		return fmt.Errorf("istiooperator MeshConfig[extensionProviders] type assertion failed: %T", obj)
+	}
+
+	for idx := range extensionProviders {
+		extensionProvider, ok := extensionProviders[idx].(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("istiooperator MeshConfig[extensionProviders][idx] type assertion failed: %T", extensionProviders[idx])
+		}
+		name, ok := extensionProvider["name"]
+		if !ok {
+			continue
+		}
+		if name == extAuthorizerName {
+			// deletes the element in the array
+			extensionProviders = append(extensionProviders[:idx], extensionProviders[idx+1:]...)
+			iop.Spec.MeshConfig["extensionProviders"] = extensionProviders
+			break
+		}
+	}
+
+	logger.Info("remove external authorizer from meshconfig")
+	if err := r.Update(ctx, iop); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *KuadrantReconciler) registerExternalAuthorizer(ctx context.Context) error {
+	logger := logr.FromContext(ctx)
+	iop := &iopv1alpha1.IstioOperator{}
+
+	iopKey := client.ObjectKey{Name: iopName(), Namespace: iopNamespace()}
+	if err := r.Get(ctx, iopKey, iop); err != nil {
+		// It should exists, NotFound also considered as error
+		logger.Error(err, "failed to get istiooperator object", "key", iopKey)
+		return err
+	}
+
+	if hasKuadrantAuthorizer(iop) {
+		return nil
+	}
+
+	//meshConfig:
+	//    extensionProviders:
+	//      - envoyExtAuthzGrpc:
+	//          port: POST
+	//          service: AUTHORINO SERVICE
+	//        name: kuadrant-authorization
+
+	if iop.Spec.MeshConfig == nil {
```

Comment:
Technically, the IstioOperator `Spec` can be `nil`. I know we control the resource, but would this panic if somebody deleted the spec and the operator tried to reconcile it back?

Reply:
nope. Spec is not a pointer in the Go struct. So it always exists

```
Spec   KuadrantSpec   `json:"spec,omitempty"`
```

Reply:
Just for the case, I checked it. 

I have deleted the spec. The k8s API serves without the spec:
```yaml
apiVersion: kuadrant.kuadrant.io/v1beta1
kind: Kuadrant
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"kuadrant.kuadrant.io/v1beta1","kind":"Kuadrant","metadata":{"annotations":{},"name":"kuadrant-sample","namespace":"default"},"spec":{}}
  creationTimestamp: "2022-05-18T13:47:47Z"
  finalizers:
    - kuadrant.kuadrant.io/finalizer
  generation: 2
  name: kuadrant-sample
  namespace: default
  resourceVersion: "5880"
  uid: 4d26b8ff-5669-48de-a9a5-5487cc079495
status:
  conditions:
    - lastTransitionTime: "2022-05-18T13:47:47Z"
      message: Kuadrant is ready
      reason: KReady
      status: "True"
      type: Ready
  observedGeneration: 2
```

But when it is read by the controller, spec object has been initialized with zero value. This is a json serialization of the go struct:
```json
{
  "kind": "Kuadrant",
  "apiVersion": "kuadrant.kuadrant.io/v1beta1",
  "metadata": {
    "name": "kuadrant-sample",
    "namespace": "default",
    "uid": "4d26b8ff-5669-48de-a9a5-5487cc079495",
    "resourceVersion": "6681",
    "generation": 2,
    "creationTimestamp": "2022-05-18T13:47:47Z",
    "annotations": {
      "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"kuadrant.kuadrant.io/v1beta1\",\"kind\":\"Kuadrant\",\"metadata\":{\"annotations\":{},\"name\":\"kuadrant-sample\",\"namespace\":\"default\"},\"spec\":{}}\n",
      "something": "jdjd"
    },
    "finalizers": [
      "kuadrant.kuadrant.io/finalizer"
    ],
  },
  "spec": {},
  "status": {
    "observedGeneration": 2,
    "conditions": [
      {
        "type": "Ready",
        "status": "True",
        "lastTransitionTime": "2022-05-18T13:47:47Z",
        "reason": "KReady",
        "message": "Kuadrant is ready"
      }
    ]
  }
}
```

Reply:
updated